test/testpushnotifications.cpp

```diff
@@ -0,0 +1,351 @@
+
+        // Right authentication data should be sent
+        QCOMPARE(spy.count(), 2);
+
+        const auto userSent = spy.at(0).at(1).toString();
+        const auto passwordSent = spy.at(1).at(1).toString();
+
+        QCOMPARE(userSent, user);
+        QCOMPARE(passwordSent, password);
+    }
+
+    void testOnWebSocketTextMessageReceived_notifyFileMessage_emitFilesChanged()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy processTextMessageSpy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(processTextMessageSpy.isValid());
+
+        auto account = createAccount().data();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account);
+        QSignalSpy filesChangedSpy(pushNotifications.get(), &OCC::PushNotifications::filesChanged);
+        QVERIFY(filesChangedSpy.isValid());
+
+        // Setup push notifications
+        pushNotifications->setup();
+
+        // Wait for authentication and then send notify_file push notification
+        processTextMessageSpy.wait();
+        QCOMPARE(processTextMessageSpy.count(), 2);
+        const auto socket = processTextMessageSpy.at(0).at(0).value<QWebSocket *>();
+        socket->sendTextMessage("notify_file");
+
+        // filesChanged signal should be emitted
+        QVERIFY(filesChangedSpy.wait());
+        QCOMPARE(filesChangedSpy.count(), 1);
+        // FIXME: This fails because accountFilesChanged is always nullptr. What do I wrong?
+        // I have checked that in a normal signal handler I get the correct account pointer.
+        // auto accountFilesChanged = filesChangedSpy.at(0).at(0).value<OCC::Account *>();
+        // QCOMPARE(accountFilesChanged, account);
+    }
+
+    void testOnWebSocketTextMessageReceived_invalidCredentialsMessage_reconnectWebSocket()
+    {
+        const QString user = "user";
+        const QString password = "password";
+        MockWebSocketServer mockServer;
+        QSignalSpy spy(&mockServer, &MockWebSocketServer::processTextMessage);
+        QVERIFY(spy.isValid());
+
+        auto account = createAccount();
+        auto credentials = new CredentialsStub(user, password);
+        account->setCredentials(credentials);
+        auto pushNotifications = createPushNotifications(account.data());
+
+        // Setup push notifications
+        pushNotifications->setup();
+
+        // Wait for authentication attempt and then sent invalid credentials
+        spy.wait();
+        QCOMPARE(spy.count(), 2);
+        const auto socket = spy.at(0).at(0).value<QWebSocket *>();
+        const auto firstPasswordSent = spy.at(1).at(1).toString();
+        QCOMPARE(firstPasswordSent, password);
+        socket->sendTextMessage("err: Invalid credentials");
+
+        // Wait for a new authentication attempt
```

Comment:
I'd clear the spy before the wait. This way you avoid having growing magic numbers in your test code.